### ISSUE TITLE
Port to libretro for fun

### ISSUE THREAD

**Issue by ghost** [LekKit/RVVM#85]
Hello I make a quick libretro port of RVVM here:
https://github.com/iyzsong/rvvm-libretro

With a modified ns16550a device to use libvterm (fb is disabled for now) for terminal output.
I think this can be useful to play some text games in RetroArch...

Is there some interest to merge a libretro port?  Thank you!

**Comment by LekKit** [comment]
As for packaging scripts needed to build it for RetroArch, I am very happy to merge them. For other stuff it is better enabled only for this target. Read more on that below.

@nebulka1 tried to bring `libvterm` support but ultimately there are some `libvterm` vulnerabilities uncovered by fuzzing (Whereas RVVM is supposed to be bulletproof). It can be all fixed but so far patches were not accepted and I am in favor of using an inhouse VT emulation (Also good for providing different VT models).

You may bring `libvterm` support for libretro, but not in default builds (because vulnerabilities in `libvterm` I am not happy to see). Just throw in some kind of `USE_VTERM=1` env to `make` and I'll handle the rest upstream soon.

Same goes for `libretro.c`, it is fine if it is only built there.

I will take some time to actually wrap my head over what's going on, but so far so good

**Comment by LekKit** [comment]
A recommendation for `libretro.c`: You do not need to explicitly create thread for `vm_run()`, just don't use blocking RVVM APIs.

The `rvvm_run_eventloop()` function is blocking, but it is not required to be called - it is more like "Wait until any machine shuts down" to know *when* they shut down. It is basically an event-based way of waiting on machines.

Therefore if you remove it and just call `rvvm_start_machine()`, it will return immediately and the VM will run. You need to move `rvvm_free_machine()` somewhere you actually destroy it, like `retro_deinit()`. Freeing a running machine is safe.

If you do not need to know when the machines shut down this is a preferred way, `threading.h` API isn't guaranteed to be stable, and you are not supposed to be using those low-level things anyways since it is prone to dataraces and such.

There is also `rvvm_on_reset` callback which does machine power events notifications though callbacks instead of event-based way.

**Comment by ghost** [comment]
Thank you for review and feedback! 

> if you remove it and just call rvvm_start_machine(), it will return immediately and the VM will run. You need to move rvvm_free_machine() somewhere you actually destroy it, like retro_deinit(). Freeing a running machine is safe.  If you do not need to know when the machines shut down this is a preferred way.

Yes, it did running fine without an extra thread, but it won't exit cleanly (random segmentfaults happend after `retro_deinit`), my guess is that `retro_deinit` does need to wait for machine shutdown / rvvm threads joined.


> there are some libvterm vulnerabilities uncovered by fuzzing.  and I am in favor of using an inhouse VT emulation.

Ah, that's bad for libvterm...  inhouse VT will be great!

Okay, I'll restruct my little code to be a RVVM fork, add libretro build option to `CMakeLists.txt`.
And modify the current `ns16550a.c` to accept different console implementations  (stdio, libvterm or maybe socket).
When it works somehow I'll open a pull request, thank you!

**Comment by LekKit** [comment]
> Yes, it did running fine without an extra thread, but it won't exit cleanly (random segmentfaults happend after `retro_deinit`), my guess is that `retro_deinit` does need to wait for machine shutdown / rvvm threads joined.

Can you please build with `CFLAGS="-g -fsanitize=address` and post a backtrace? The `rvvm_free_machine()` is supposed to cleanly kill a machine if it's running (It calls `rvvm_pause_machine()` internally). Maybe there is a bug somewhere, I should know

> And modify the current ns16550a.c to accept different console implementations (stdio, libvterm or maybe socket).

Yeah, great. Thank you too!

**Comment by ghost** [comment]
> Can you please build with CFLAGS="-g -fsanitize=address and post a backtrace? The rvvm_free_machine() is supposed to cleanly kill a machine if it's running (It calls rvvm_pause_machine() internally). Maybe there is a bug somewhere, I should know

Well I unable to get a meaningful stacktrace, maybe due to lacking debug symbols of glibc..
```
Continuing.
AddressSanitizer:DEADLYSIGNAL
=================================================================
==13619==ERROR: AddressSanitizer: SEGV on unknown address 0x7fffebd27264 (pc 0x7fffebd27264 bp 0x608000000228 sp 0x7fffe514f960 T3)
==13619==The signal is caused by a READ memory access.

Thread 4 "retroarch" received signal SIGSEGV, Segmentation fault.
0x00007ffff44b4770 in ?? () from /gnu/store/094bbaq6glba86h1d4cj16xhdi6fk2jl-gcc-10.3.0-lib/lib/libgcc_s.so.1
(gdb) bt
#0  0x00007ffff44b4770 in ?? () from /gnu/store/094bbaq6glba86h1d4cj16xhdi6fk2jl-gcc-10.3.0-lib/lib/libgcc_s.so.1
#1  0x00007ffff44b626b in _Unwind_Backtrace () from /gnu/store/094bbaq6glba86h1d4cj16xhdi6fk2jl-gcc-10.3.0-lib/lib/libgcc_s.so.1
#2  0x00007ffff74d4510 in ?? () from /gnu/store/094bbaq6glba86h1d4cj16xhdi6fk2jl-gcc-10.3.0-lib/lib/libasan.so.6
#3  0x00007ffff74b7e4d in ?? () from /gnu/store/094bbaq6glba86h1d4cj16xhdi6fk2jl-gcc-10.3.0-lib/lib/libasan.so.6
#4  0x00007ffff74cc9e0 in ?? () from /gnu/store/094bbaq6glba86h1d4cj16xhdi6fk2jl-gcc-10.3.0-lib/lib/libasan.so.6
#5  0x00007ffff74b540c in ?? () from /gnu/store/094bbaq6glba86h1d4cj16xhdi6fk2jl-gcc-10.3.0-lib/lib/libasan.so.6
#6  0x00007ffff74b3226 in ?? () from /gnu/store/094bbaq6glba86h1d4cj16xhdi6fk2jl-gcc-10.3.0-lib/lib/libasan.so.6
#7  0x00007ffff74b266e in ?? () from /gnu/store/094bbaq6glba86h1d4cj16xhdi6fk2jl-gcc-10.3.0-lib/lib/libasan.so.6
#8  <signal handler called>
#9  0x00007fffebd27264 in ?? ()
#10 0x0000000000002ad9 in ?? ()
#11 0x000000003102675f in ?? ()
#12 0x0000000000000000 in ?? ()
```

And once I got a stacktrace  in `builtin_eventloop` after adding some printf statement but forget to save it, now unable to reproduce it...

My guess is that after `thread_detach`, there is still some memory access going in those RVVM threads, and after `retro_deinit` retroarch will dlclose the core (it loads rvvm as a shared library), making those threads's memory address invalid.

**Comment by LekKit** [comment]
> My guess is that after `thread_detach`, there is still some memory access going in those RVVM threads, and after `retro_deinit` retroarch will dlclose the core (it loads rvvm as a shared library), making those threads's memory address invalid.

Ahh it makes sense now. I will try to reproduce on my side, maybe RVVM is not dlclose-friendly.

**Comment by fish4terrisa-MSDSM** [comment]
> Thank you for review and feedback!
> 
> > if you remove it and just call rvvm_start_machine(), it will return immediately and the VM will run. You need to move rvvm_free_machine() somewhere you actually destroy it, like retro_deinit(). Freeing a running machine is safe.  If you do not need to know when the machines shut down this is a preferred way.
> 
> Yes, it did running fine without an extra thread, but it won't exit cleanly (random segmentfaults happend after `retro_deinit`), my guess is that `retro_deinit` does need to wait for machine shutdown / rvvm threads joined.
> 
> > there are some libvterm vulnerabilities uncovered by fuzzing.  and I am in favor of using an inhouse VT emulation.
> 
> Ah, that's bad for libvterm... inhouse VT will be great!
> 
> Okay, I'll restruct my little code to be a RVVM fork, add libretro build option to `CMakeLists.txt`. And modify the current `ns16550a.c` to accept different console implementations (stdio, libvterm or maybe socket). When it works somehow I'll open a pull request, thank you!

the socket method is a good idea as i can connect it using socat.libvterm may not be supported for Android,but socat + socket is available for Android.

**Comment by ghost** [comment]
> the socket method is a good idea as i can connect it using socat.libvterm may not be supported for Android,but socat + socket is available for Android.

@fish4terrisa-MSDSM Okay, I'll try to see what I can get...


> And modify the current ns16550a.c to accept different console implementations (stdio, libvterm or maybe socket).

@LekKit I have some ideas about it, could you help me to see if I'm on the right road?

1. Add a `rvvm_console_t` struct for backends for ns16550a:
```
typedef struct {
    ssize_t (*read)(void *buf, size_t nbytes);
    ssize_t (*write)(const void *buf, size_t nbytes);
    int (*ready)(void);
} rvvm_console_t;

extern rvvm_console_t stdio_console;

PUBLIC void ns16550a_init(rvvm_machine_t* machine, rvvm_console_t* console,
                          rvvm_addr_t base_addr, plic_ctx_t* plic, uint32_t irq);
PUBLIC void ns16550a_init_auto(rvvm_machine_t* machine, rvvm_console_t* console);
```
`read` and `write` just like their POSIX ones, `ready` will return `0` for not ready (eg: socket server but no client connected), `1` for ready to read, `2` for ready to write, `3` for ready to both read and write.

2. Change `ns16550a_data` struct to:
```
struct ns16550a_data {
    plic_ctx_t* plic;
    uint32_t irq;
    spinlock_t lock;

    uint8_t ier;
    uint8_t iir;
    uint8_t lcr;
    uint8_t mcr;
    uint8_t lsr;
    uint8_t msr;
    uint8_t scr;
    uint8_t dll;
    uint8_t dlm;

    rvvm_console_t* console;
};
```

`iir`, `lsr` and `msr` added, the goal is to implement more realistic interrupts behaviour.

I find a reference here:
https://marknelson.us/posts/1990/05/01/servicing-com-port-interrupts.html

IIUC,
- in ns16550a_update, check console for read/write ready, update LSR, if LSR updated (we care RECV 0x01 and THR 0x60) also update IIR, and then send irqs according to IER and IIR.
- in ns16550a_mmio_read, when read LSR or MSR, I clear its bits in IIR, which means the interrupts have been handled in the guest.
- When console returns error (eg: EOF) for read/write, clear the Carrier Detect in MSR, so an irq will be fired in `ns16550a_update` soon.


How about it?  Thank you!

**Comment by LekKit** [comment]
See https://github.com/LekKit/RVVM/issues/86#issuecomment-1501665872

So yeah, I was basically working on similar thing but it is not upstreamed and I'm waiting for your PR in any way you like it, then I'll refactor it - that way we'll prevent merge conflicts and responsibility for the API lies on me then. I hope you get the rationale.

> @fish4terrisa-MSDSM Okay, I'll try to see what I can get...

Assuming they meant TCP sockets and not UNIX ones, you may reuse the portable `networking.h` lib (Cover under #ifdef USE_NET), or I may simply add it later.

> `iir`, `lsr` and `msr` added, the goal is to implement more realistic interrupts behaviour.

You are free to do so, but perhaps split that in a separate PR so I will do additional tests if this works properly with all guests, code review. We have problems with OpenBSD guests (It waits for some characters to be pressed before doing more output). If you can fix that, would be awesome.

Reproduce: U-Boot from https://github.com/LekKit/patches-misc/releases/tag/rvvm-uboot-2023.4, OpenBSD from https://cdn.openbsd.org/pub/OpenBSD/snapshots/riscv64/miniroot73.img

**Comment by ghost** [comment]
Closed this issue now, new development in #96
